### ISSUE TITLE
Inlined function after switch/case not semantically equivalent

### ISSUE THREAD

**Issue by jfo84** [terser/terser#244]
**Bug report or Feature request?**

Bug

**Version (complete output of `terser -V`)**

3.14.1

**Complete CLI command or `minify()` options used**

terser --compress --mangle -- input.js

**`terser` input**

```
			var l = 2 * s,
				p = t.x,
				h = t.y,
				d = function (e, t, r, n, o)
				{
					var i = function (e)
					{
						switch (e)
						{
						case "left":
							return {
								arrowX: -1,
								arrowY: 0
							};
						case "right":
							return {
								arrowX: 1,
								arrowY: 0
							};
						case "top":
							return {
								arrowX: 0,
								arrowY: -1
							};
						case "bottom":
							return {
								arrowX: 0,
								arrowY: 1
							};
						default:
							return {
								arrowX: 0,
								arrowY: 0
							}
						}
					}(o);
					return {
						xe: e + i.arrowX * r * n / 2,
						ye: t + i.arrowY * r * n / 2
					}
				}(n.x, n.y, l, c, i),
				y = d.xe,
				v = d.ye,
				b = function (e, t, r, n, o)
				{
					return "top" === o || "bottom" === o ?
					{
						xa1: e,
						ya1: t + (n - t) / 2
					} : "left" === o || "right" === o ?
					{
						xa1: e + (r - e) / 2,
						ya1: t
					} :
					{
						xa1: e,
						ya1: t
					}
				}(p, h, y, v, r),
```

**`terser` output or error**

```
				var l = 2 * s,
					p = t.x,
					h = t.y,
					d = function (e, t, r, n, o)
					{
						var i = function (e)
						{
							switch (o)
							{
							case "left":
								return {
									arrowX: -1,
									arrowY: 0
								};
							case "right":
								return {
									arrowX: 1,
									arrowY: 0
								};
							case "top":
								return {
									arrowX: 0,
									arrowY: -1
								};
							case "bottom":
								return {
									arrowX: 0,
									arrowY: 1
								};
							default:
								return {
									arrowX: 0,
									arrowY: 0
								}
							}
						}();
						return {
							xe: e + i.arrowX * r * n / 2,
							ye: t + i.arrowY * r * n / 2
						}
					}(n.x, n.y, l, c, i),
					y = d.xe,
					v = d.ye,
					b = function (e, t, r, n, o)
					{
						return "top" === o || "bottom" === o ?
						{
							xa1: p,
							ya1: h + (v - h) / 2
						} : "left" === o || "right" === o ?
						{
							xa1: p + (y - p) / 2,
							ya1: h
						} :
						{
							xa1: p,
							ya1: h
						}
					}(0, 0, 0, 0, r),
```

**Expected result**

The original code or a semantically equivalent version. The original code has already gone through TerserWebpackPlugin (internally using 3.10.12). I used Terser itself for generating the example.

**Comment by fabiosantoscode** [comment]
Is this the smallest code that produces the bug? Could you please try to make it as minimal as possible?

**Comment by jfo84** [comment]
@fabiosantoscode I can try to generate something smaller this week.

At first glance, it looks like two bugs. This is what would happen if the default case in the switch/case is always assumed to happen. Then, this code is assumed to always evaluate to 0 for all cases:

```js
{
  xe: e + i.arrowX * r * n / 2,
  ye: t + i.arrowY * r * n / 2
}
``

It doesn't feel like that's a valid assumption either, since if `e` or `t` is non-zero you can't optimize the entire expression away.

**Comment by jfo84** [comment]
After looking at this again, the bug seems to be related to destructuring an object directly returned from a function.

The original code looks like this:
```js
const { arrowX, arrowY } = computeEndingArrowDirectionVector(endingAnchor);
```

where `computeEndingArrowDirectionVector` is the above switch/case

**Comment by fabiosantoscode** [comment]
Thanks for your comments. So what exactly is a code that can reproduce this? Do you mean that computeEndingArrowDirectionVector is a function that contains all the original code that was posted?

**Comment by jfo84** [comment]
I haven't been able to reduce it to anything below a one thousand-line dump; the initial post is an excerpt of that where the problem occurred

I question the usefulness of the heuristic in general and think it would be difficult to generate a real test case if you don't have any ideas as to why this is happening, which is concerning on its own

**Comment by fabiosantoscode** [comment]
It's very possible to generate a real test case by removing code until you have a minimal piece of code which reproduces the issue.

Even though I understand it's a bit hard to do this, as I'm sure you'll understand I can't fix a problem if it can't be reproduced.

**Comment by jridgewell** [comment]
I looked at this again, and I'm not able to reproduce any issue. Maybe this has been fixed in the 2 years since being filed?

More perplexing, the OP's input/output code are equivalent. I think a lot of focus is on the `}(0, 0, 0, 0, r),` in the last line, but this isn't actually an error. The input code has `}(p, h, y, v, r),`, and those variable are referenced as the IIFE's params `(e, t, r, n, o)`. Terser took each reference to `e`/`t`/`r`/`n` in the IIFE body and remapped it to the argument's `p`/`h`/`y`/`v`. Confusingly, it left the function's params instead of dropping them, but the overall minification is sound.

I think this should be closed as not reproducible.
